File: windows-native/NexaLink.Tests/NexaLinkTests.cs
using System;
using NexaLink.Helpers;
using NexaLink.Services;

namespace NexaLink.Tests
{
    /// <summary>
    /// Unit tests for Windows native app.
    /// Run with: dotnet test
    /// </summary>
    public class NexaLinkTests
    {
        // ===== Utils =====

        public static void TestFormatSize()
        {
            Assert(Utils.FormatSize(0) == "0 B");
            Assert(Utils.FormatSize(512) == "512 B");
            Assert(Utils.FormatSize(1024) == "1 KB");
            Assert(Utils.FormatSize(1048576) == "1 MB");
            Console.WriteLine("✅ FormatSize");
        }

        public static void TestMatrixUsername()
        {
            Assert(Utils.MatrixUsername("@admin:server.com") == "admin");
            Assert(Utils.MatrixUsername("@user123:matrix.org") == "user123");
            Console.WriteLine("✅ MatrixUsername");
        }

        public static void TestMatrixServer()
        {
            Assert(Utils.MatrixServer("@admin:server.com") == "server.com");
            Assert(Utils.MatrixServer("@admin:server.com:8448") == "server.com:8448");
            Console.WriteLine("✅ MatrixServer");
        }

        public static void TestInitials()
        {
            Assert(Utils.Initials("John Doe") == "JD");
            Assert(Utils.Initials("Admin") == "AD");
            Assert(Utils.Initials("A") == "A");
            Console.WriteLine("✅ Initials");
        }

        public static void TestGetMimeType()
        {
            Assert(Utils.GetMimeType("photo.jpg") == "image/jpeg");
            Assert(Utils.GetMimeType("video.mp4") == "video/mp4");
            Assert(Utils.GetMimeType("song.mp3") == "audio/mpeg");
            Assert(Utils.GetMimeType("doc.pdf") == "application/pdf");
            Assert(Utils.GetMimeType("file.xyz") == "application/octet-stream");
            Console.WriteLine("✅ GetMimeType");
        }

        // ===== MatrixClient =====

        public static void TestMxcToHttp()
        {
            var client = new MatrixClient("https://example.com");
            Assert(client.MxcToHttp("mxc://example.com/abc123") == "https://example.com/_matrix/media/v3/download/example.com/abc123");
            Assert(client.MxcToHttp(null) == null);
            Assert(client.MxcToHttp("https://not-mxc") == null);
            Assert(client.MxcToHttp("mxc://noslash") == null);
            Console.WriteLine("✅ MxcToHttp");
        }

        // ===== Models =====

        public static void TestRoomModel()
        {
            var room = new Models.RoomModel { RoomId = "!abc", Name = "Test Room", LastMessageTime = 1700000000000 };
            Assert(room.Initials == "TE");
            Assert(room.TimeDisplay != "");
            Console.WriteLine("✅ RoomModel");
        }

        public static void TestMessageModel()
        {
            var msg = new Models.MessageModel { EventId = "e1", Sender = "@admin:server", Body = "Hello", MsgType = "m.text", Timestamp = 1700000000000 };
            Assert(msg.SenderName == "admin");
            Assert(msg.TypeIcon == "");
            var imgMsg = new Models.MessageModel { MsgType = "m.image" };
            Assert(imgMsg.TypeIcon == "📷");
            var audioMsg = new Models.MessageModel { MsgType = "m.audio" };
            Assert(audioMsg.TypeIcon == "🎤");
            Console.WriteLine("✅ MessageModel");
        }

        // ===== LoginResponse =====

        public static void TestLoginResponse()
        {
            var success = new LoginResponse { UserId = "@user:s", AccessToken = "token", DeviceId = "DEV" };
            Assert(success.AccessToken != null);
            Assert(success.ErrorCode == null);
            var error = new LoginResponse { ErrorCode = "M_FORBIDDEN", Error = "Bad password" };
            Assert(error.AccessToken == null);
            Assert(error.ErrorCode == "M_FORBIDDEN");
            Console.WriteLine("✅ LoginResponse");
        }

        // ===== Encryption =====

        public static void TestEncryptedPayload()
        {
            var payload = new EncryptedPayload { Ciphertext = "ct", EncryptedKey = "ek", IV = "iv" };
            Assert(payload.Ciphertext == "ct");
            Assert(payload.EncryptedKey == "ek");
            Assert(payload.IV == "iv");
            Console.WriteLine("✅ EncryptedPayload");
        }

        // ===== OfflineQueue =====

        public static void TestQueuedMessage()
        {
            var msg = new OfflineQueue.QueuedMessage { RoomId = "!room", Body = "Hello", Retries = 3 };
            Assert(msg.Retries == 3);
            Assert(msg.Body == "Hello");
            Console.WriteLine("✅ QueuedMessage");
        }

        // ===== RelativeTime =====

        public static void TestRelativeTime()
        {
            var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            Assert(Utils.RelativeTime(now) == "just now");
            Assert(Utils.RelativeTime(now - 300000).Contains("m ago"));
            Console.WriteLine("✅ RelativeTime");
        }

        // ===== Run all =====

        public static void RunAll()
        {
            Console.WriteLine("=== NexaLink Windows Tests ===");
            TestFormatSize();
            TestMatrixUsername();
            TestMatrixServer();
            TestInitials();
            TestGetMimeType();
            TestMxcToHttp();
            TestRoomModel();
            TestMessageModel();
            TestLoginResponse();
            TestEncryptedPayload();
            TestQueuedMessage();
            TestRelativeTime();
            Console.WriteLine("=== All tests passed! ===");
        }

        private static void Assert(bool condition)
        {
            if (!condition) throw new Exception("Assertion failed");
        }
    }
}

File: android-native/app/src/test/java/io/nexalink/app/NexaLinkTests.kt
package io.nexalink.app

import io.nexalink.app.network.MatrixApi
import io.nexalink.app.network.OfflineQueue
import org.junit.Assert.*
import org.junit.Test

/**
 * Unit tests for core NexaLink components.
 */
class MatrixApiTest {

    @Test
    fun mxcToHttp_validUrl() {
        val api = MatrixApi("https://example.com")
        val result = api.mxcToHttp("mxc://example.com/abc123")
        assertEquals("https://example.com/_matrix/media/v3/download/example.com/abc123", result)
    }

    @Test
    fun mxcToHttp_invalidUrl() {
        val api = MatrixApi("https://example.com")
        assertNull(api.mxcToHttp("https://not-mxc"))
        assertNull(api.mxcToHttp(null))
        assertNull(api.mxcToHttp(""))
    }

    @Test
    fun mxcToHttp_noSlash() {
        val api = MatrixApi("https://example.com")
        assertNull(api.mxcToHttp("mxc://noslash"))
    }
}

class LoginResponseTest {

    @Test
    fun loginResponse_success() {
        val resp = io.nexalink.app.network.LoginResponse(
            user_id = "@user:server",
            access_token = "token123",
            device_id = "DEVICE1"
        )
        assertNotNull(resp.access_token)
        assertNotNull(resp.user_id)
        assertNull(resp.errcode)
    }

    @Test
    fun loginResponse_error() {
        val resp = io.nexalink.app.network.LoginResponse(
            errcode = "M_FORBIDDEN",
            error = "Invalid password"
        )
        assertNull(resp.access_token)
        assertNotNull(resp.errcode)
    }
}

class MessageEntityTest {

    @Test
    fun messageEntity_creation() {
        val msg = io.nexalink.app.data.MessageEntity(
            eventId = "evt1",
            roomId = "!room:server",
            sender = "@user:server",
            body = "Hello",
            msgtype = "m.text",
            timestamp = 1000L
        )
        assertEquals("evt1", msg.eventId)
        assertEquals("Hello", msg.body)
        assertEquals("m.text", msg.msgtype)
    }

    @Test
    fun messageEntity_mediaTypes() {
        val types = listOf("m.text", "m.image", "m.video", "m.audio", "m.file")
        for (type in types) {
            val msg = io.nexalink.app.data.MessageEntity(
                eventId = "e", roomId = "r", sender = "s",
                body = "test", msgtype = type, timestamp = 0
            )
            assertEquals(type, msg.msgtype)
        }
    }
}

class RoomEntityTest {

    @Test
    fun roomEntity_creation() {
        val room = io.nexalink.app.data.RoomEntity(
            roomId = "!room:server",
            name = "Test Room",
            lastMessage = "Hello",
            lastMessageTime = 1000L
        )
        assertEquals("!room:server", room.roomId)
        assertEquals("Test Room", room.name)
    }

    @Test
    fun roomEntity_defaults() {
        val room = io.nexalink.app.data.RoomEntity(
            roomId = "!room:server",
            name = "Room"
        )
        assertNull(room.avatarUrl)
        assertNull(room.lastMessage)
        assertEquals(0, room.lastMessageTime)
        assertEquals(0, room.unreadCount)
        assertFalse(room.isDirect)
    }
}

class EncryptionTest {

    @Test
    fun encryptedPayload_creation() {
        val payload = io.nexalink.app.network.EncryptedPayload(
            ciphertext = "abc",
            encryptedKey = "def",
            iv = "ghi"
        )
        assertEquals("abc", payload.ciphertext)
        assertEquals("def", payload.encryptedKey)
        assertEquals("ghi", payload.iv)
    }

    @Test
    fun megolmPayload_creation() {
        val payload = io.nexalink.app.network.MegolmPayload(
            sessionId = "sid",
            ciphertext = "ct",
            iv = "iv",
            mac = "mac",
            messageIndex = 5
        )
        assertEquals("sid", payload.sessionId)
        assertEquals(5, payload.messageIndex)
    }
}

class MediaCacheEntityTest {

    @Test
    fun mediaCacheEntity_creation() {
        val entry = io.nexalink.app.data.MediaCacheEntity(
            mxcUrl = "mxc://server/media1",
            localPath = "/data/media/file.jpg",
            mimeType = "image/jpeg",
            size = 1024
        )
        assertEquals("mxc://server/media1", entry.mxcUrl)
        assertEquals("/data/media/file.jpg", entry.localPath)
        assertEquals(1024, entry.size)
    }
}

class UtilsTest {

    @Test
    fun userIdParsing() {
        val userId = "@admin:72.56.244.207"
        val name = userId.split(":")[0].removePrefix("@")
        val server = userId.split(":")[1]
        assertEquals("admin", name)
        assertEquals("72.56.244.207", server)
    }

    @Test
    fun roomAlias() {
        val alias = "#nexalink-shorts-v3:72.56.244.207"
        assertTrue(alias.startsWith("#"))
        assertTrue(alias.contains(":"))
    }

    @Test
    fun mimeTypeDetection() {
        val types = mapOf(
            "photo.jpg" to "image/jpeg",
            "video.mp4" to "video/mp4",
            "audio.ogg" to "audio/ogg",
            "doc.pdf" to "application/pdf"
        )
        for ((file, expected) in types) {
            val ext = file.substringAfterLast(".")
            val mime = when (ext) {
                "jpg", "jpeg" -> "image/jpeg"
                "png" -> "image/png"
                "mp4" -> "video/mp4"
                "ogg" -> "audio/ogg"
                "pdf" -> "application/pdf"
                else -> "application/octet-stream"
            }
            assertEquals(expected, mime)
        }
    }
}
